### PR TITLE
perf: adding native UDP socket (NanoSockets) for supported platforms

### PR DESCRIPTION
**Pull request by phodoval**

### CODE REVIEW

**Review by Lymdun — APPROVED** [review]

**Comment by James-Frowen** [comment]
test running here https://github.com/MirageNet/Mirage/runs/3079229937

File: Assets/Mirage/Runtime/Sockets/Udp/NanoSocket.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+        public void Send(IEndPoint endPoint, byte[] packet, int length)
+        {
+            var nanoEndPoint = (NanoEndPoint)endPoint;
+            UDP.Send(socket, ref nanoEndPoint.address, packet, length);
+        }
+
+        void InitSocket()
+        {
+            socket = UDP.Create(256 * 1024, 256 * 1024);
```
Comment (Lymdun) [inline Assets/Mirage/Runtime/Sockets/Udp/NanoSocket.cs:99]: Maybe buffer size should be configurable?

File: Assets/Mirage/Runtime/Sockets/Udp/UdpSocketFactory.cs [hunk 2]
```diff
@@ -2,40 +2,85 @@
 using System.Net;
 using System.Net.Sockets;
 using Mirage.SocketLayer;
+using NanoSockets;
 using UnityEngine;
 
 namespace Mirage.Sockets.Udp
 {
+    public enum SocketLib { Automatic, NanoSockets, Net };
```
Comment (Lymdun) [inline Assets/Mirage/Runtime/Sockets/Udp/UdpSocketFactory.cs:10]: Maybe rename NanoSockets to Native and Net to Managed? Might be easier to understand

File: Assets/Mirage/Runtime/Sockets/Udp/NanoSockets/Scripts/NanoSockets.cs [hunk 3]
```diff
@@ -0,0 +1,222 @@
+/*
```
Comment (Lymdun) [inline Assets/Mirage/Runtime/Sockets/Udp/NanoSockets/Scripts/NanoSockets.cs:1]: Shouldn't we use git submodules instead than copying the file manually?

File: Assets/Mirage/Runtime/Sockets/Udp/NanoSocket.cs (hunk 1 above)
Comment (phodoval) [reply Assets/Mirage/Runtime/Sockets/Udp/NanoSocket.cs]: Done

File: Assets/Mirage/Runtime/Sockets/Udp/UdpSocketFactory.cs (hunk 2 above)
Comment (phodoval) [reply Assets/Mirage/Runtime/Sockets/Udp/UdpSocketFactory.cs]: Done

File: Assets/Mirage/Runtime/Sockets/Udp/NanoSockets/Scripts/NanoSockets.cs (hunk 3 above)
Comment (phodoval) [reply Assets/Mirage/Runtime/Sockets/Udp/NanoSockets/Scripts/NanoSockets.cs:1]: Adding a submodule to the repo will probably require some CI tweaks to also fetch the sumbodules when checking out etc. The same would go for anyone who would want to clone the repo. Also not sure about UPM. I do use the git submodules personally but I must say it's quite tedious.

**Pull request merged**